### ISSUE TITLE
Using better transformer and ROPE dynamic configuration results in useless qloras

### ISSUE THREAD

**Issue by DiegoSVilla** [huggingface/trl#938]
Hey! I have been doing some qlora finetunes using trl import SFTTrainer, DataCollatorForCompletionOnlyLM (on llama v2).

My problem starts with max length 8k for rope finetuning crashes my 4090 even on 7bmodel,  4 bit, small loras , the solution I found online was using bettertransformer and that solved it, but its not compatible with pad tokens aparently, and DataCollatorForCompletionOnlyLM cant have packing true:

"The BetterTransformer implementation does not support padding during training, as the fused kernels do not support attention masks. Beware that passing padded batched data during training may result in unexpected outputs. Please refer to https://huggingface.co/docs/optimum/bettertransformer/overview for more details."

Still, using this with linear rope configuration works ok, A problem I had is the output trying to generate 2 or 3 times the response if it was too short. Here I think the fixed 8k length when I have input prompts varying from 1k tokens to 8k starts to impact.

Now, when I try dynamic setting for ROPE then it just becomes a useless model. 

My recomendation here I think is a VariableLengthDataset, that will at least separate some granularity of lengths, 64,128,256....up to max 8k, so it would be slower for sure, but would better handle dynamic rope scaling and maybe minimyze padding inconsistencies with bettertransformers?

If my script / model config are relevant for the case, I will be thriled to share!

Also, if the solution is already out there and I didnt find it please point me to it!

**Comment by matrixssy** [comment]
+1
